larky/src/main/java/com/verygood/security/larky/modules/AccountUpdaterModule.java

```diff
@@ -0,0 +1,123 @@
+import com.verygood.security.larky.modules.vgs.aus.spi.AccountUpdaterService;
+import java.util.List;
+import java.util.ServiceLoader;
+import net.starlark.java.annot.Param;
+import net.starlark.java.annot.ParamType;
+import net.starlark.java.annot.StarlarkBuiltin;
+import net.starlark.java.annot.StarlarkMethod;
+import net.starlark.java.eval.*;
+
+@StarlarkBuiltin(
+    name = "native_au",
+    category = "BUILTIN",
+    doc = "Overridable Account Updater API in Larky")
+public class AccountUpdaterModule implements LarkyAccountUpdater {
+  public static final AccountUpdaterModule INSTANCE = new AccountUpdaterModule();
+
+  public static final String ENABLE_MOCK_PROPERTY =
+      "larky.modules.vgs.nts.enableMockAccountUpdater";
+
+  private final AccountUpdaterService AccountUpdaterService;
+
+  public AccountUpdaterModule() {
+    final ServiceLoader<AccountUpdaterService> loader =
+        ServiceLoader.load(AccountUpdaterService.class);
+    final List<AccountUpdaterService> AccountUpdaterProviders =
+        ImmutableList.copyOf(loader.iterator());
+
+    if (Boolean.getBoolean(ENABLE_MOCK_PROPERTY)) {
+      AccountUpdaterService = new MockAccountUpdaterService();
+    } else if (AccountUpdaterProviders.isEmpty()) {
+      AccountUpdaterService = new NoopAccountUpdaterService();
+    } else {
+      if (AccountUpdaterProviders.size() != 1) {
+        throw new IllegalArgumentException(
+            String.format(
+                "AccountUpdaterModule expecting only 1 network token provider of type AccountUpdaterService, found %d",
+                AccountUpdaterProviders.size()));
+      }
+      AccountUpdaterService = AccountUpdaterProviders.get(0);
+    }
+  }
+
+  @StarlarkMethod(
+      name = "lookup_card",
+      doc = "Lookup account updates for a given PAN.",
+      useStarlarkThread = true,
+      parameters = {
+        @Param(
+            name = "number",
+            named = true,
+            doc = "Card PAN. Used to look up the corresponding card information to be returned",
+            allowedTypes = {@ParamType(type = String.class)}),
+        @Param(
+            name = "exp_month",
+            named = true,
+            doc =
+                "Card expiration month. Used to pass to the network for retrieving the corresponding card information",
+            allowedTypes = {@ParamType(type = StarlarkInt.class)}),
+        @Param(
+            name = "exp_year",
+            named = true,
+            doc =
+                "Card expiration year. Used to pass to the network for retrieving the corresponding card information",
+            allowedTypes = {@ParamType(type = StarlarkInt.class)}),
+        @Param(
+            name = "name",
+            named = true,
+            doc =
+                "Card owner name. Used to pass to the network for retrieving the corresponding card information",
+            allowedTypes = {@ParamType(type = String.class)}),
+        @Param(
```

Comment:
For me, this immediately stood out as 'hmm.' as the credentials are passed in via method call. Is this what we are intending?

Reply:
Yes, that's what we are intending.

This is for our needs in payments team to make it possible for our customer to get the latest card number from account updater service (provided by visa or mastercard). And to achieve that, we need to make a call to public API of calm. I envisioned this could be used like this

```python
if use_account_updater(input.headers):
    card = aus.lookup_card(
        pan=body.cardNumber,
        exp_month=body.expMon,
        exp_year=body.expYear,
        name=body.name,
        client_id="<CALM SERVICE ACCOUNT CLIENT ID>",
        client_secret="<CALM SERVICE ACCOUNT CLIENT SECRET>",
    )
    if card is not None and card.number != input.cardNumber or ...:
        input.cardNumber = card.number
        # update card
# proceed to PSP with or without new card info
```

As you can see the challenge here is to authenticate with calm API server. The easiest way right now is to hardwire the service account credentials in the larky script. I know this may not be the best way in the world to authenticate to our own API service in VGS, but considering we need this feature in short time, it's the best way. If there's a concern for hardwiring the credentials in the larky script, we can actually ask the merchant to put their service account credentials in the vault and use `reveal` function to get the actual credential secret values. Like this


```python
card = aus.lookup_card(
     # ...
    client_id=vault.reveal("<CALM SERVICE ACCOUNT CLIENT ID>"),
    client_secret=vault.reveal("<CALM SERVICE ACCOUNT CLIENT SECRET>"),
)
```

And if we don't want the possibility of merchant putting their service account credentials in larky script and force them to use vault, we can even reveal the secret in module and reject the request if the provided values are not alias.

In a perfect world, if we have enough time to build this feature and some major changes to vault is allowed, I was thinking about providing something like OIDC auth token like the ones from Kubernetes pod for authenticating with Calm server. Since the larky script environment is kind of already a trusted environment, I don't see why we need to authenticate again. So requests coming from larky can be trusted as long as there's a policy says that we trust calm API requests coming from our own vault proxy. Then the API call can be simplified as just 

```python
card = aus.lookup_card(
    pan=body.cardNumber,
    exp_month=body.expMon,
    exp_year=body.expYear,
    name=body.name,
)
```

The account updater module will create a OIDC auth token for making the request and calm can authenticate and accept that automatically based on a pre-defined policy.

Regardless, that's too much effort for now, as we want this feature to be done ASAP, provides if keeping calm's service account credentials in vault (or even hardwiring in larky script) is acceptable, I think this is the best approach for now.

Reply:
To me the ability to use account updater should be done through a module. I don't feel that we should be using credentials like this for service to service calls. Here is my thinking.

Create a Vault Preference called "Account Updater" that is default false.
If that preference is false - horizon, simply does not allow the call to the service be completed.
If that preference is true - it goes through

The preference can only be set by someone/something with permissions to do so. ( this is something that the new Vault Management spec intends to support)

Reply:
so it sounds like we only want to enable this feature for customers who really need it, is my understanding correct?

is this vault preference feature something we can use right now, or is this a new feature and not yet implemented? I did a quick search in the code base and cannot find something looks like vault preference. If it's already there, can you please point that to me where's it and if there's an example of it?

Reply:
Yes, it's there... @sharifelgamal or @Iapetus999 can point you in the right direction. We made it so that preferences for the vault are available in horizon. Adding a new preference - 🤷🏼  not sure about. but I am going to hope it can be quick to add one🤞🏼 .

Reply:
note that we never actually added the ability to read preferences from horizon, it would involve the proxy (or operation pipeline as it were) passing in the preference as a header into larky